### ISSUE TITLE
Allow ufunc inner loops to access dtypes

### ISSUE THREAD

**Issue by eric-wieser** [numpy/numpy#12518]
Mentioned in a bunch of places, but would be valuable to have a single place to track this.

To be able to reuse an inner loop across multiple dtypes (ie across variable-length strings, structured arrays, etc), we need access to the `dtype` object within the inner loop.

Perhaps we can achieve this by adding a new `npy_bool innerloop_takes_dtype` flag to ufuncs, which when set indicates that the inner loop signature now also takes a `PyArray_Descr **dtypes` argument.

**Comment by eric-wieser** [comment]
> There was consensus a few years ago about hiding the ufunc struct internals from the C API

This gets a +1 from me. I hadn't considered the fact that we can wrap one into the other.

Another option... In C, is it safe to call a function pointer with more arguments that it expects? Won't it just put extra items on its stack that it can ignore?

**Comment by seberg** [comment]
Is there consensus that this is the right approach? Because my gut feeling is that it might not be.

Basically, what operation would you want the dtype for in a more generic loop where it is not better to do it either before or after the all of the inner operations? My gut feeling would rather be:

 1. Run dtype specific Loop Selection/Preparation function before actual calculations:
     * This can error out (say a unit error)
     * It could in principle even return the inner loop
     * It would set/create the new dtype object.
     * It can return a `void *` object to be passed in as additional data to the ufunc, which could be, but does not have to be the dtype.
 2. Possibly run another function after the inner loop is finshed (for example to allow error handling, since the inner loop should be able to unlock the GIL).

I find it a bit strange to specifically pass in the `dtype` object to the inner loop, because I think we need a "setup" step anyway, and that setup step can decide what it wants to pass in. Of course for some "generic" inner loops (wrapping python objects or so), they may end up passing in the dtype, but that would be implementation.

Not sure I make sense, but it seems to me that almost all logic derived from the dtype inside the inner loop would rather already be handled before calling the inner loop.

**Comment by mhvk** [comment]
I'm also not sure we want to pass on `dtype` by default, since in general the inner loops should be simple and fast, and not have to deal with details like it. Even for structured dtypes, I'm not so sure I understand the general case; even for equality, if it weren't for being unable to override the `ufunc`, the current outside implementation would seem the proper solution (which suggests we just need to move some of the logic from `ndarray.__eq__` to one of the setup functions in the `np.equal` ufunc).

For strings, the case seems clearer, but it does seem that the machinery is in principle present to deal with it: we should be able to use the `void *` for passing on, e.g., length information.

All this said, it does seem worth checking exactly where those helper functions get called, so that we can make sure, e.g., it is possible to return a custom ufunc loop or to create an output array with a custom dtype.

**Comment by njsmith** [comment]
We probably should rework the ufunc loop APIs in general to support setup/teardown/dispatch in a more logical way. (For example, np.add on strings will need some way to compute the output dtype before the actual loop starts running.)

But we should still pass the dtype information to every loop. Forcing people to jump through hoops to access this basic information makes no sense to me. Writing np.equal on strings shouldn't require writing a special setup trampoline that just copies a pointer from one place to another.

Possible new ufunc loop signature:

* nin
* nout
* dtypes (array of dtype object pointers with nin+nout entries)
* data (array of char*'s with nin+nout entries)
* outer_count: how many times the core operation should be performed (single integer)
* outer_strides: how far to step between each core operation (array of integers with nin+nout entries)
* core_ndims: the dimensionality of each core argument (array of integers with nin+nout entries, all zero for scalar ufuncs)
* core_shapes: the shapes of each core argument (use ndims to decode; can be NULL if core_ndims is all-zero)
* core_strides: the strides within each core operation (same layout as core_shapes)
* `void *per_call_user_data`: arbitrary data passed from the loop setup function
* `void *per_loop_user_data`: arbitrary data passed when registering the loop function.

**Comment by njsmith** [comment]
Here's a use case that might clarify things. At some point we'll want to expose the loop machinery to python, so people can write loops implemented in python, for prototyping and for all the stuff people use `vectorize` for.

So: given a proposal for the C loop signature, how hard would it be to write a *single* C loop function that can be registered with *any* ufunc, and turns around and invokes a python function instead? Obviously this is much easier if the loop has information on dtypes, so it can convert the raw data pointers into ndarrays to pass to python.

**Comment by seberg** [comment]
Good point, all flexible dtypes will need it in some form or another, that is definitely true. It is just, for downstream it seems to me that improving setup/teardown is more important. For the support of flexible dtypes, passing the dtypes would be the most convenient thing for sure.

So I am not against it, it will hardly hurt and is indeed more convenient in many cases. Too many flags (or typically unused args) may also not be convenient though, so before that happens, I am not sure if a minimal solution may not be just as well. But, we will need an iteration on the inner loop API anyway probably, so there is probably no point to trying to make due with the current signature.

I am sure it is already on everyone's list, but one other API change is that a stop iteration/error integer return value may be a good idea.

**Comment by seberg** [comment]
@njsmith well, but you need to write a single C-loop+setup func, I think we could live with that single loop /setup func being inconvenient to write :).

**Comment by seberg** [comment]
Anyway, I am not really against it. I just thought it might be focusing on the wrong thing. But that is probably just me not being up to the full picture.

**Comment by seberg** [comment]
Anyway, I am not against it. I just thought it might be focusing on the wrong thing. But that is probably just me not being up to the full picture.

**Comment by njsmith** [comment]
> I am sure it is already on everyone's list, but one other API change is that a stop iteration/error integer return value may be a good idea.

Good point! Loops can raise exceptions already, so we might as well make things less weird and have them return a -1 to signal that, like every other python C API. And it would let us stop manually checking for exceptions after every call.

Re: error handing, the other thing to think about is the floating point error flags. Currently we clear the flags before every call, and check them afterwards. This is pretty expensive on some CPUs, so it would be nice to be able to skip it for things like integer or string operations where it's irrelevant.

I guess the simplest way to do this would be to declare that new-style loops are responsible for doing their own checking for floating point errors, and providing a simple API to help them do this (e.g. call `PyArray_FloatErr_Start` at the beginning of your loop, and `PyArray_FloatErr_Check` at the end).

**Comment by njsmith** [comment]
Oh man, there's also the GIL handling to think about. What even is the current behavior? Something like, loops are called with the GIL dropped, unless there's a special flag set on one of the dtypes?

**Comment by shoyer** [comment]
This is probably a good time for me to share my notes (still in proto-NEP form) on defining high level dtypes/ufuncs from Python: https://hackmd.io/6YmDt_PgSVORRNRxHyPaNQ?both

**Comment by seberg** [comment]
Yeah, it is based on dtype flags (on the iterator), I think some of them might be set becaues e.g. casting can cause arrays. It might be nice to be able to force GIL for specific ufuncs.

@shoyer thanks for sharing. Just glancing over, I am wondering if you guys discussed an API that is more setup/teardown related. So, instead of actually returning a result, I think `__dtype_ufunc__` should have to return `add(double, double)` inner loop and `dtype=double[meters]` as output dtype.

I would expect a python defined "fast" inner loop (for 1-D arrays), to be forced to use `out[...] = res` to be honest. Similar as what would be necessary in nditer.

**Comment by shoyer** [comment]
@seberg Yeah, that's the natural alternative, which I think could also work, e.g.,

```python
def sub_dates(x, y, out):
    np.sub(x.view(np.int64), y.view(np.int64), out=out.view(np.int64)

class MyDatetime(metaclass=np.dtype):
    @classmethod
    def __dtype_ufunc__(cls, ufunc, method, *input_dtypes, **kwargs):
        if method != '__call__':
            return NotImplemented
        if ufunc is np.sub:
            a, b = inputs_dtypes
            if issubclass(a, cls) and issubclass(b, cls):
                return sub_dates, MyTimedelta
            elif issubclass(b, MyTimedelta):
                return sub_dates, MyDatetime
            else:
                return NotImplemented
        # implement other ufuncs
        return NotImplemented
```

Throwing out some pros/cons:

- Advantages of `__dtype_ufunc__` returning an inner loop and a dtype:
    - It's more similar to what NumPy already does.
    - NumPy will setup the outputs with the proper shapes/dtypes, so less validation is necessary.
    - We can restrict the the dtype loop generator from knowing about array values by only giving it access to dtypes, which restricts users from writing bad ufuncs.
- Advantages of `__dtype_ufunc__` implementing the ufunc:
    - Slightly easier for users to understand/implement, since there is one less layer of indirection.
    - More similar to `__array_ufunc__`.
    - For some (poorly designed?) ufuncs, you might not know the right dtype until you look over the data, e.g., for concatenating two fixed length strings you might want the result dtype to be the smallest length possible.

Anything I missed here?

**Comment by shoyer** [comment]
Thinking about this a little more, returning an inner loop and a dtype from `__dtype_ufunc__` (or whatever we call it, maybe `__array_ufunc_resolver__`?) does seem like a much better idea, even though I do have a slight dislike for impure functions. I will think about how to rewrite the rest of the high level API proposal around that.

**Comment by njsmith** [comment]
I changed the issue title, since the conversation has broadened a bit, and since we probably only want to change the calling convention once, so keeping the conversation together seems more useful than splitting it up.

**Comment by shoyer** [comment]
Coming back on topic: for high level APIs that define a ufunc inner loop from Python, we definitely need the low-level calling convention to support:

1. Constructing actual NumPy arrays with of the given dtypes for inputs and `out`.
2. Calling an arbitrary Python callable on these arrays.

**Comment by njsmith** [comment]
@shoyer The whole `__dtype_ufunc__` thing seems odd to me. Right now, we have a mechanism for extending ufuncs to handle dtypes, and it's not this – it's the multi-method-style dispatch that ufuncs do, by maintaining a table of loops that they choose between. I'm not a big fan of multi-method dispatch in general, but in this particular case it has history behind it, and also solves a real problem: a single call might involve a user-defined ufunc, multiple user-defined dtypes, and in fact the dtypes we end up using in the call might not even appear in the inputs (because of casting). So there's no obvious place to put the dunder method that can handle all the different extension points.

(I also thing it might be too ambitious to try to wedge *all* dtype-specific functions into the ufunc mold. Obviously junk like `dot` and `clip` should become ufuncs instead of staying on the dtype, but dtypes do have a core set of operations like casting, copying, and boxing/unboxing that are more primitive than ufuncs, and that the ufunc machinery needs to use to do its job – I'd make these dedicated methods on the dtype object.)

Anyway, back to the ufunc/loop API: my thought for making this more Python-friendly would be to keep the current architecture, but expose a Python API for it. So, we'd have a `UFunc` object, which has-a table of `UFuncLoop` objects, and each `UFuncLoop` has metadata like which dtypetypes it can handle, the actual setup/loop/teardown functions, etc. And we'd have Python-level APIs for adding/removing `UFuncLoop`s from a `UFunc`, introspecting the metadata, etc.

**Comment by eric-wieser** [comment]
To be clear here: My goal was much shorter-term than providing python support - I wanted something minimal here to let `argmin` and `sort` ufuncs have access to `f->compare` so that we can eliminate the sorting functions from the dtypes and sort.

I don't think we want to wait for a full picture of python-level overrides before we pull a lot of the ufuncable stuff out of #12514, and this looks like the first step we need to achieve in some way.

**Comment by njsmith** [comment]
Yeah, supporting parametric dtypes properly definitely creates some gnarly details to sort out :-). And I haven't thought it all through either.

But the basic argument for needing a dispatch table in the ufunc object is very simple. Suppose you're writing a third-party library and want to define a custom ufunc that works on floats, like `scipy.special.erf`. Presumably you aren't going to go mutate numpy's float64 dtype to make its `__dtype_ufunc__` method know about `scipy.special.erf`. So the ufunc has to have a dispatch table in it, at least for built-in types. And if you want to avoid having "two different notions of dtype", then that should be how *all* dispatch is done, rather than using one dispatch mechanism for built-in dtypes and a different one for user-defined dtypes.

Also, what if I want to define my own ufunc that works on datetime64?

> I guess this is where @seberg's "Run dtype specific Loop Selection/Preparation function" comes in: once we've identified that we can cast arguments to string/string, now we ask string to tell us what the exact (fixed width) result dtype is?

Well, probably you don't want to ask string, probably you want to ask the loop object representing string/string addition. But yeah, basically. You can also imagine precomputing some things (e.g. for `equal` on a categorical dtype, you might build a table mapping between the two sets of categories?), or handling other stuff (maybe we should allow ufuncs to take kwargs, so that `sort` could be a ufunc, and then the preparation function could somehow use the kwargs to decide which loop to use?). We don't need to figure out those details right now... the basic intuition though is that there will clearly have to be *some* code that runs before the actual loop to figure out the output dtypes, so it's reasonable to assume that we'll want some way to pass data from that code into the loop function, so the loop signature should have a `void*` reserved for that use.

> The dispatch table does a nice job of solving the use-case of being able to combine two dtypes A and B that know nothing about each other, if they can be cast to dtypes for which we have an inner loop defined. That's great in theory, but I'm not sure how often it is necessary in practice. Probably never, with NumPy's current dtypes

We actually do use this a lot right now, e.g. most ufuncs don't know about `float16`, but it works anyway thanks to casting:

```
In [1]: scipy.special.erf.types                                                 
Out[1]: ['f->f', 'd->d', 'F->F', 'D->D']

In [2]: scipy.special.erf(np.array(1.0, dtype=np.float16))                      
Out[2]: 0.8427008
```

Anyway, my first attempt would be: well, we already need a distinction between what I've been calling "a dtype" (an instance object with state, e.g. "string of length 7"), and "a dtypetype" (a type object, subclass of `np.dtype`, e.g. "fixed length string"). We obviously don't want to define seperate ufunc loops for different lengths of strings, so that suggests that the ufunc loop dispatch table should be organized by dtypetype. (This is also consistent with how method lookup happens in Python in general.) And then given a set of concrete input dtypes, we'll need a way for each loop to compute the concrete output dtype. (An invariant I hope we can have: *which* loop we choose depends on the inputs, not the outputs, and only their dtypetypes, not their dtypes. The actual computation of course will depend on the dtypes. Also if someone passes in an explicit `out` array, I'm not sure if we want to let the loop see the concrete dtype, or if we should let it do its normal thing and then cast if necessary. Though we'll at least need an equality operation on dtypes to figure out whether we need to allocate a temporary array and then cast, or whether we can use the given output array directly.)

You're right that it would be annoying to have to explicitly say "yeah, this output float64 is exactly like every other float64, because float64 is a singleton dtypetype". (Well, there's some question about how to handle endian-swapping, which actually *is* a way our float64 dtypes can vary right now, but maybe we can handle that as a wrapper dtype that takes some other dtype and swaps the bytes, or something.) Also, our backcompat wrappers are going to need to be able to cope with being given a loop function without an attached setup function. But maybe this is just a matter of having a pre-defined setup function that you just pass by rote whenever you're using boring built-in types? Or we could have a concept of "default construction" for dtypetypes.

(BTW, casting has some closely related subtleties: in some cases we want to cast data to a particular dtype, e.g. `existing_array[:] = array_of_a_different_dtype`, and in other cases we want to cast data to a particular dtypetype and let the cast function choose the concrete dtype, like casting an array of strings to categorical. If we use something like my scheme above, then ufunc casting will use )

> linear search over the dispatch table gets slower with every new dtype.

Linear search is AFAICT unavoidable if we want to allow casting, because there's no way to know a priori which things can be cast to what, except to ask them. But there are a lot of things we can do to speed this up. If there's an exact match, you can find that in O(1) and only fall back to linear search if there isn't. Once you find a match via linear search you can memoize it and now it's an exact match. (Maybe you have to reset the memo cache whenever a loop is added/removed, but that's an extremely rare operation.)

This does have some implications for the casting and ufunc design though. In particular, when selecting a loop, we *could* take into account the full dtype of the inputs – we already know what they are! – but if we restrict ourselves to only taking into account the dtypetype, then that makes memoization much more effective. So this also pushes towards having dtypetype-level "can you cast?" operations, and using those exclusively for ufunc selection.

**Comment by shoyer** [comment]
> But the basic argument for needing a dispatch table in the ufunc object is very simple. Suppose you're writing a third-party library and want to define a custom ufunc that works on floats, like `scipy.special.erf`. Presumably you aren't going to go mutate numpy's float64 dtype to make its `__dtype_ufunc__` method know about `scipy.special.erf`. So the ufunc has to have a dispatch table in it, at least for built-in types. And if you want to avoid having "two different notions of dtype", then that should be how _all_ dispatch is done, rather than using one dispatch mechanism for built-in dtypes and a different one for user-defined dtypes.

This may be the fundamental design tradeoff. Either:

1. We have two different types of ufunc dispatch: "for dtypetype" and "for dtype given a dtypetype"
2. We have two different types of ufunc dispatch: defined on ufuncs and defined on dtypes.

I don't really *like* either of this options but it seems like we'll have to pick. Either way, in practice we will be able to reserve the second type of dispatch for unusual cases.

For what it's worth, I agree that all things being equal putting dispatch in ufuncs makes more sense. I would try to save dispatch on dispatch on dtypes only for new dtypes / old ufuncs.

The biggest virtue of supporting defining dispatch on dtypes might be that you could potentially define generic implementations of every ufunc, e.g., by casting to another dtype and then casting back. For example, this would probably be the easiest way to implement a variable length string dtype in NumPy, by doing all computation on object dtype arrays. (That said, it would probably work nearly as well to expose a generic override for implementing the inner loop of "every built-in arithmetic operation".)

> Also, what if I want to define my own ufunc that works on datetime64?

You would have to define casting rules for handling inputs with different dtypes. The difference is that you would do this by providing a "default dtype resolver" function, rather than by specifying a dtypetype implementation + "dtypetype specific dtype resolver".

datetime64 is actually an easy case: you can rely on NumPy's builtin casting rules for datetime64/timedelta64 to give the same units for everything, e.g., by making use of a protocol version of `np.result_type`. Or if you're pandas, you can only bother to implement your functions for datetime64[ns].

I guess a virtue of "dispatching on ufuncs" is that you can get this sort of casting for free, so much of the time you don't need to write the "dtypetype specific dtype resolver." But it's probably not that much more difficult to make a generic "dtype resolver that uses casting."

Cases like new ufuncs for a custom units dtype will be tricky either way.

**Comment by njsmith** [comment]
> We have two different types of ufunc dispatch: "for dtypetype" and "for dtype given a dtypetype"

Sorry, maybe I'm being tired, but I don't know what these two types are :-). Can you explain?

**Comment by shoyer** [comment]
> > We have two different types of ufunc dispatch: "for dtypetype" and "for dtype given a dtypetype"
> 
> Sorry, maybe I'm being tired, but I don't know what these two types are :-). Can you explain?

"for dtypetype": This is what currently have in NumPy, i.e., the loops associated with each ufunc. These can't handle metadata.

"for dtype given a dtypetype": This is the new thing we would need, to pick the specific dtypes/loop (e.g., `(string[2], string[3]) -> string[5]`) based on the dtypetypes (`(string, string) -> string)`) and the ufunc (`np.add`)

**Comment by njsmith** [comment]
The "dispatch" part (i.e., picking which loop to use) is identical between those: in both cases you look at the input [1] dtypetypes and find the loop that matches (possibly after casting).

The difference comes after you pick which loop you want to use, and only depends on the output dtypetypes. If one or more of those dtypetypes requires some parameters to instantiate, then the loop needs a mechanism to specify those parameters. OTOH if all the output dtypetypes don't require any parameters (they're effectively singletons), then this mechanism collapses into triviality.

Calling this "two different types of ufunc dispatch" feels weird to me, because one is a special case of the other, and anyway, the difference happens after the dispatch phase :-). And anyway... the reason we're forced to add the concept of a "loop setup function" is to handle output dtypedtypes that require parameters to instantiate, but once we have the concept, it probably has other uses, even for non-parametric dtypetypes (e.g. precomputing things).

[1] I think in some cases currently, if an `out=` argument is given, we might also take its dtype into account when picking which loop to use, which means that passing `out=` arguments can change the output. If we can get rid of this I think that would be good.

**Comment by shoyer** [comment]
> The "dispatch" part (i.e., picking which loop to use) is identical between those: in both cases you look at the input [1] dtypetypes and find the loop that matches (possibly after casting).

I'm not sure we can pick loops merely based on dtypetype. Some parametric dtypes might correspond to different inner loops even with the same dtypetype, e.g., consider dtypes like Masked or Quantity that could wrap any number of low-level dtypes.

I guess this might not be a meaningful distinction if the loop has access to dtype metadata, e.g., so it can call the ufunc again on NumPy arrays if necessary.

**Comment by njsmith** [comment]
I was going to reply to your first paragraph, but your second paragraph already said what I was going to say :-)

Though actually, this does suggest: maybe our ufunc inner loop signature should also receive the ufunc object? Similar to how methods receive `self`?

**Comment by mattip** [comment]
Maybe we need a pure python implementation we can play with to see how this works. It should be simple without too many decision points. 

One of the detriments of the current mechanism(s) is that the decisions are made far away from where they are used, making debugging difficult. Being able to stop in a debugger and walk back up a linear stack to see how choices were made is a definite plus.

**Comment by shoyer** [comment]
> Though actually, this does suggest: maybe our ufunc inner loop signature should also receive the ufunc object? Similar to how methods receive `self`?

Sure, though I suppose you could also easily put this into `void *per_loop_user_data` if desired.

In practice I think most new dtypes will be implemented via high level Python interfaces (if possible), so it doesn't matter as much what we do for the lower level C interface.

**Comment by eric-wieser** [comment]
> Maybe we need a pure python implementation we can play with to see how this works.

It seems to me like we're trying to design a fully-fledged function-overloading system here, and trying to cover many of the parametric cases that C++ templates would cover. This seems like it might be in peril of falling outside the scope of numpy, although I'll admit that custom dtypes lose significant value without it.

**Comment by seberg** [comment]
Yes, we are designing a fully-fledged function-overloading system in a sense. I doubt we can avoid that, I am not even quite sure how much we already have? It would be nice to keep it very simple. I think it might be ok to remove implicit upcasting (do not cast "ee->e" to "ff->f", since there is no overlap at all; we can add/wrap the loops manually for backcompat).

Considering the use case of making progress with flexible types, maybe we can side step most of this discussion if we can settle the signature change.

The idea of a self could be neat. For the python side it seems convenient, but that does not mean it has to exist on the C-side. I am not sure there is much of a use case on the C-side, since the inner loop should be simple and usually unique to a single loop.

**Comment by njsmith** [comment]
Though... arrays with missing values might be a problem here, now that I think about it :-/

For units, it's not a big deal, because units behave differently in different operations (e.g. add has to check unit compatibility, multiply creates new units, etc.), and part of defining a units package is making a list of which operations you want to support and writing code to support each of them. So that fits well with the model of attaching loops to each of the ufuncs you want to support.

But for missing values, you want to essentially support *all* ufuncs that have support for the wrapped dtype. And the details are complicated – I'm not sure a simple `__dtype_ufunc__` even solves it, once you consider cases like a ufunc that takes two arguments and both of them are passed as different missing-value dtypes, and that if the inputs are missing-value dtypes then the output has to be too.

I think the ufunc machinery will have to have some hard-coded knowledge of missing values in any case (e.g., we'll want ufuncs to support a generic `skipNA=` kwarg). Maybe the thing to do is worry about these details later, when figuring out missing value extensions specifically, rather than trying to make the basic dtype machinery generic enough to handle missing values as a special case.

**Comment by seberg** [comment]
I have started to work on a draft here https://github.com/numpy/numpy/pull/16723

The current plan is much like some of the posts above, although details around how to "reuse" loops best seem a bit complicated to me.

1. Promotion: I am tending towards multiple-dispatching like promotion, we could also have a list of promoters, etc. but you need some way to "order" them probably.  The biggest advantage of proper/multiple promoters, is that a Promoter could dynamically create new loops.

2. For the general loop signature, I think we can limit ourselves to "strided" loops for now, possibly with allowing loop specialization (e.g. if one of the strides is 0 or contiguous; which are things that casting currently does).  I think the normal (and likely only) inital way should be to just give a single strided loop though.

3. On the actual *signature* change:
   * I think it is clear we want a return value. I am considering using a nonzero return to break the loop. A *negative* return value would indicate that a Python error is already set.
   * I find it hard to give a reasonable set of things that should be passed in, so I am considering to rather pass in a structure with the following fields:
      1. ``ufunc``
      2. ``descriptors *``
      3. ``threadstate``  (so loops can safely set an error without dancing around it in setup/teardown)
      4. ``static loop data`` (current ``void *``, or maybe some other Python object)
      5. ``void *per-run loop data``
      6. "``self``"?  I would not call it a true self, this is not really an instance or class. But there is some information, such as the ``identity`` which lives on the ``UFuncImpl``/Loop.  Some notion of "self" (or at least additionally passed in object) probably needs to exist for Python, and it being special to some degree may be necessary for easier refcounting.

   As you can see that list is *long* and I am not even sure that there may not be something I missed.  A struct at least could grow in theory.  Plus, many numeric loops actually do not need *any* of the information (because we can handle the floating point error management), so this means they do not have to list 5-6 unused parameters.
   * I think we need setup/teardown as I mentioned before. But, I am leaning towards a simple independent flag for floating point error handling (it is pretty special after all).  In fact, while I think we should have setup/teardown, and should expose it eventually.  I am happy not to rush that.

The signature decisions have overlap with casting (although casting does not have a clear "ufunc" that can be passed in).  Casting is almost a ufunc, but not quite (possibly they could use the same but just pass ``NULL`` for ``ufunc``, and reconsider later).

**Comment by mattip** [comment]
I think most of this has been incorporated into NEP 43 which was merged (in draft form).
